pkg/controllers/provisioning/v1alpha1/reallocation/terminate.go

```diff
@@ -19,24 +19,115 @@ import (
 	"fmt"
 	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
 	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/utils/functional"
+	"github.com/awslabs/karpenter/pkg/utils/ptr"
+	"github.com/awslabs/karpenter/pkg/utils/scheduling"
 	"go.uber.org/zap"
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 type Terminator struct {
 	kubeClient    client.Client
 	cloudprovider cloudprovider.Factory
+	coreV1Client  corev1.CoreV1Interface
 }
 
-// DeleteNodes will use a cloudprovider implementation to delete a set of nodes
-func (t *Terminator) DeleteNodes(ctx context.Context, nodes []*v1.Node, spec *v1alpha1.ProvisionerSpec) error {
-	// 1. Delete node in cloudprovider's instanceprovider
-	if err := t.cloudprovider.CapacityFor(spec).Delete(ctx, nodes); err != nil {
-		return fmt.Errorf("terminating %d cloudprovider instances, %w", len(nodes), err)
+func (t *Terminator) Reconcile(ctx context.Context, provisioner *v1alpha1.Provisioner) error {
+	// 1. Cordon terminable nodes
+	if err := t.cordonNodes(ctx, provisioner); err != nil {
+		return fmt.Errorf("cordoning terminable nodes, %w", err)
+	}
+
+	// 2. Drain and delete nodes
+	if err := t.terminateNodes(ctx, provisioner); err != nil {
+		return fmt.Errorf("terminating nodes, %w", err)
 	}
+	return nil
+}
 
-	// 2. Delete node in APIServer to ensure
+// cordonNodes takes in a list of expired nodes as input and cordons them
+func (t *Terminator) cordonNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) error {
+	// 1. Get terminable nodes
+	nodeList, err := t.getNodes(ctx, provisioner, map[string]string{
+		v1alpha1.ProvisionerPhaseLabel: "terminable",
```

Comment:
We should keep track of these states in the v1alpha1 package, since we hardcode the string in multiple places. 

type LifecyclePhase string
const (
  LifecyclePhaseTerminable LifecyclePhase = "Terminable"
)

Reply:
Ah yes. That's a great idea. Thanks!